colossalai/inference/tensor_parallel/engine.py

```diff
@@ -59,17 +115,45 @@ def __init__(self,
 
         self.dtype = dtype
 
-        self.head_dim = model.config.hidden_size // model.config.num_attention_heads
-        self.head_num = model.config.num_attention_heads
-        self.layer_num = model.config.num_hidden_layers
+        self.head_dim = self.model.config.hidden_size // self.model.config.num_attention_heads
+        self.head_num = self.model.config.num_attention_heads
+        self.layer_num = self.model.config.num_hidden_layers
 
-        self.tp_size = -1    # to be set with given shard config in self.prepare_shard_config
         self.cache_manager = None
 
-        self.shard_config = shard_config
-        self.model = None
-        # optimize the original model by sharding with ShardFormer
-        self._optimize_model(model=model.to(device))
+    def _get_model_and_tokenizer(self, model: str, tokenizer: str, trust_remote_code: bool) -> nn.Module:
+
+        if isinstance(model, nn.Module) and isinstance(tokenizer, (PreTrainedTokenizer, PreTrainedTokenizerFast)):
+            return model, tokenizer
+
+        supported_model = model
+
+        try:
+            config = AutoConfig.from_pretrained(model, trust_remote_code=trust_remote_code)
+        except ValueError as e:
+            if (not trust_remote_code and "requires you to execute the configuration file" in str(e)):
+                err_msg = ("Failed to load the model config. If the model is a custom "
+                           "model not yet available in the HuggingFace transformers "
+                           "library, consider setting `trust_remote_code=True` in LLM "
+                           "or using the `--trust-remote-code` flag in the CLI.")
+                raise RuntimeError(err_msg) from e
+            else:
+                raise e
+        architectures = getattr(config, "architectures", [])
+        for arch in architectures:
+            if arch in _supported_models:
+                if isinstance(tokenizer, str):
+                    if arch == "LlamaForCausalLM" or arch == "LLaMAForCausalLM" or arch == "LlamaModel":
+                        tokenizer = LlamaTokenizer.from_pretrained(model, trust_remote_code=trust_remote_code)
+                    else:
+                        tokenizer = AutoTokenizer.from_pretrained(model, trust_remote_code=trust_remote_code)
+                if isinstance(model, str):
+                    supported_model = _supported_models[arch].from_pretrained(model,
+                                                                              pad_token_id=tokenizer.eos_token_id)
+
+                return supported_model, tokenizer
+        raise ValueError(f"Model architectures {architectures} are not supported for now. "
+                         f"Supported architectures: {list(_supported_models.keys())}")
```

Comment:
cleaner way: replace `list(_supported_models.keys())` with `self.supported_models` (<--we have this property)

Reply:
Done